README.md
# RobustAdaptiveMetropolisSampler

[![Project Status: WIP – Initial development is in progress, but there has not yet been a stable, usable release suitable for the public.](https://www.repostatus.org/badges/latest/wip.svg)](https://www.repostatus.org/#wip)
[![Build Status](https://travis-ci.org/anthofflab/RobustAdaptiveMetropolisSampler.jl.svg?branch=master)](https://travis-ci.org/anthofflab/RobustAdaptiveMetropolisSampler.jl)
[![Build status](https://ci.appveyor.com/api/projects/status/66r0lf8kcim6fm0o/branch/master?svg=true)](https://ci.appveyor.com/project/anthofflab/robustadaptivemetropolissampler-jl/branch/master)
[![codecov](https://codecov.io/gh/anthofflab/RobustAdaptiveMetropolisSampler.jl/branch/master/graph/badge.svg)](https://codecov.io/gh/anthofflab/RobustAdaptiveMetropolisSampler.jl)

## Overview

This package implements the robust adaptive metropolis (RAM) sampler described in [Vihola (2012)](https://doi.org/10.1007/s11222-011-9269-5) for the [Julia language](https://www.julialang.org).

## Usage

The `RAM_sample` function runs a MCMC sampler on a given log target function. The arguments for the functions are as follows:

```julia
RAM_sample(logtarget, x0, M0, n; opt_α=0.234, γ=2/3, q=Normal(), show_progress=true)
```

* `logtarget` this must be a callable that accepts one parameter which is a vector of values to evaluate the log target function on. The function passed must return the log value of the target function.
* `x0` is a vector of initial values at which the sampler will start the MCMC algorithm. The length of the vector controls the dimensionality of the problem.
* `M0` is the initial co-variance matrix that the sampler should use to scale the new proposal. `M0` can be passed in many different ways:
1) a scalar: an isotropic covariance matrix with diagonal elements `abs2(M0)`.
2) an `AbstractVector`: a diagonal covariance matrix with diagonal elements `abs2.(M0)`.
3) an `AbstractMatrix` (or a `Diagnoal` or an `AbstractPDMat`): a value of any of these types will be interpreted directly as the covariance matrix.
* `n`: the number of elements to be sampled, i.e. the length of the chain.
* `opt_α`: the target acceptance rate the algorithm is trying to hit.
* `γ`: a parameter for the computation of the step size sequence.
* `q`: the proposal distribution.
* `show_progress`: a flag that controls whether a progress bar is shown.

The function returns a `NamedTuple` with three elements:
* `chain`: a `Matrix` with the result chain. Each row is one sample, the columns correspond to the dimensions of the problem.
* `acceptance_rate`: the acceptance rate for the overall chain.
* `M`: the last co-variance matrix used in the algorithm.

A simple example of using the function is

```julia
using Distributions, RobustAdaptiveMetropolisSampler

chain, accrate, S = RAM_sample(
    p -> logpdf(Normal(3., 2), p[1]), # log target function
    [0.],                             # Initial value
    0.5,                              # Use an isotropic covariance matrix with diagonal elements abs2(0.5)
    100_000                           # Number of runs
)
```


NEWS.md
# RobustAdaptiveMetropolisSampler.jl v1.0.0 Release Notes
* Fix incorrect log target comparison

# RobustAdaptiveMetropolisSampler.jl v0.2.0 Release Notes
* API changes

# RobustAdaptiveMetropolisSampler.jl v0.1.0 Release Notes
* Initial release


test/runtests.jl
using RobustAdaptiveMetropolisSampler
using Distributions
using Statistics
using LinearAlgebra
using Random
using Test

@testset "RobustAdaptiveMetropolisSampler" begin

Random.seed!(10)

result = RAM_sample(p->logpdf(Normal(3.,2.), p[1]), [0.], 0.5, 100_000, show_progress=false)

@test length(result.chain) == 100_000
@test result.acceptance_rate ≈ 0.234 atol=0.01
@test mean(result.chain[:,1]) ≈ 3. atol=0.1
@test std(result.chain[:,1]) ≈ 2. atol=0.01

result = RAM_sample(p->logpdf(Normal(3.,2.), p[1]), [0.], [0.5], 1_000, show_progress=false)
@test length(result.chain) == 1_000

result = RAM_sample(p->logpdf(Normal(3.,2.), p[1]), [0.], Diagonal([0.5]), 1_000, show_progress=false)
@test length(result.chain) == 1_000


result = RAM_sample(p->logpdf(Normal(3.,2.), p[1]), [0.], fill(0.5, 1, 1), 1_000, show_progress=false)
@test length(result.chain) == 1_000

end


src/MCMCRAMSampler.jl
module MCMCRAMSampler

using LinearAlgebra, Random, Distributions, PDMats

export RAM_sample

# TODO This should be removed once the PR in PDMats with this is merged
LinearAlgebra.cholesky(a::PDiagMat) = cholesky(Diagonal(a.diag))
LinearAlgebra.cholesky(a::ScalMat) = cholesky(Diagonal(fill(a.value, a.dim)))

function RAM_sample(logtarget, x0::AbstractVector{<:Number}, s0::Matrix{<:Real}, n::Int; kwargs...)
    return RAM_sample(logtarget, x0, PDMat(s0), n; kwargs...)
end

function RAM_sample(logtarget, x0::AbstractVector{<:Number}, s0::Vector{<:Real}, n::Int; kwargs...)
    return RAM_sample(logtarget, x0, PDiagMat(abs2.(s0)), n; kwargs...)
end

function RAM_sample(logtarget, x0::AbstractVector{<:Number}, s0::Diagonal{<:Real}, n::Int; kwargs...)
    return RAM_sample(logtarget, x0, PDiagMat(diag(s0)), n; kwargs...)
end

function RAM_sample(logtarget, x0::AbstractVector{<:Number}, s0::Real, n::Int; kwargs...)
    return RAM_sample(logtarget, x0, ScalMat(length(x0), abs2(s0)), n; kwargs...)
end

function RAM_sample(logtarget, x0::AbstractVector{<:Number}, s0::AbstractPDMat, n::Int; opt_α=0.234, γ=2/3, q=Normal())
    n > 0 || error("n must be larger than 0.")
    0 < opt_α < 1 || error("opt_α must be between 0 and 1.")
    0.5 < γ <= 1 || error("γ must be between 0.5 and 1.")

    x = copy(x0)
    s = cholesky(s0)

    d = length(x0)

    u = zeros(d)
    y = zeros(d)

    log_probability_x = logtarget(x0)

    output_chain = Matrix{Float64}(undef, n, d)

    stats_accepted_values = 0

    for i in 1:n
        # Step R1
        rand!(q, u)
        y[:] .= x .+ s.L * u

        # Step R2
        log_probability_y = logtarget(y)

        α = min(1, exp(log_probability_y - log_probability_x))

        if α > rand()
            stats_accepted_values += 1
            x, y = y, x
        end

        # Step R3

        # This is taken from the second paragraph of section 5
        η = min(1, d * i^-γ)

        M = s.L * (I + η * (α-opt_α) * (u * u') / norm(u)^2 ) * s.L'
        # The paper has a proof that M is symmetric, so we declare that fact
        # to work around numerical rounding errors
        s = cholesky(Symmetric(M))

        output_chain[i, :] .= x
    end

    return (chain=output_chain, acceptance_rate=stats_accepted_values/n, S=s.L)
end

end


src/RobustAdaptiveMetropolisSampler.jl
module RobustAdaptiveMetropolisSampler

using LinearAlgebra, Random, Distributions, PDMats, ProgressMeter

export RAM_sample

# The following methods cover different ways to pass in a co-variance matrix
function RAM_sample(logtarget, x0::AbstractVector{<:Number}, M0::AbstractMatrix{<:Real}, n::Int; kwargs...)
    return RAM_sample(logtarget, x0, PDMat(M0), n; kwargs...)
end

function RAM_sample(logtarget, x0::AbstractVector{<:Number}, M0::AbstractVector{<:Real}, n::Int; kwargs...)
    return RAM_sample(logtarget, x0, PDiagMat(abs2.(M0)), n; kwargs...)
end

function RAM_sample(logtarget, x0::AbstractVector{<:Number}, M0::Diagonal{<:Real}, n::Int; kwargs...)
    return RAM_sample(logtarget, x0, PDiagMat(diag(M0)), n; kwargs...)
end

function RAM_sample(logtarget, x0::AbstractVector{<:Number}, M0::Real, n::Int; kwargs...)
    return RAM_sample(logtarget, x0, ScalMat(length(x0), abs2(M0)), n; kwargs...)
end

# Actual sampling code

function RAM_sample(logtarget, x0::AbstractVector{<:Number}, M0::AbstractPDMat, n::Int; opt_α=0.234, γ=2/3, q=Normal(), show_progress::Bool=true)
    length(x0) == size(M0, 1) || error("Covariance matrix M0 must match size of x0.")
    n > 0 || error("n must be larger than 0.")
    0 < opt_α < 1 || error("opt_α must be between 0 and 1.")
    0.5 < γ <= 1 || error("γ must be between 0.5 and 1.")

    x = copy(x0)
    s = cholesky(M0)

    d = length(x0)

    u = zeros(d)
    y = zeros(d)

    log_probability_x = logtarget(x0)

    output_chain = Matrix{Float64}(undef, n, d)

    stats_accepted_values = 0

    progress_meter = Progress(n, show_progress ? 0.1 : Inf)

    # This is a pre-allocated vector used in the loop
    scaled_proposal_vector = Vector{Float64}(undef, d)

    for i in 1:n
        # Step R1
        rand!(q, u)
        
        y[:] .= x .+ mul!(scaled_proposal_vector, s.L, u)

        # Step R2
        log_probability_y = logtarget(y)

        α = min(1, exp(log_probability_y - log_probability_x))

        if α > rand()
            stats_accepted_values += 1
            x, y = y, x
            log_probability_x = log_probability_y
        end

        # Step R3

        # This is taken from the second paragraph of section 5
        η = min(1, d * i^-γ)

        # Compute the new covariance matrix
        M = s.L * (I + η * (α-opt_α) * (u * u') / norm(u)^2 ) * s.L'

        # The paper has a proof that M is symmetric, so we declare that fact
        # to work around numerical rounding errors
        s = cholesky(Symmetric(M))

        output_chain[i, :] .= x

        next!(progress_meter; showvalues = [(:acceptance_rate,stats_accepted_values/i)])
    end

    return (chain=output_chain, acceptance_rate=stats_accepted_values/n, M=s.L * s.L')
end

end
